src/token-providing-subscription.ts

```diff
@@ -1,147 +1,161 @@
-                                    {
-                                        onOpen: headers => {
-                                            onTransition(new OpenSubscriptionState(headers, this.underlyingSubscription, onTransition));
-                                        },
-                                        onRetrying: listeners.onRetrying,
-                                        onError: error => {
-                                            if(isTokenExpiredError(error)){
-                                                tokenProvider.clearToken(token);
-                                                fetchTokenAndExecuteSubscription();
-                                            }
-                                            else{
-                                                onTransition(new FailedSubscriptionState(error));
-                                            }
-                                        },
-                                        onEvent: listeners.onEvent,
-                                        onEnd: error => {
-                                            onTransition(new EndedSubscriptionState(error));
-                                        }
-                                    },
-                                    headers
-                                )
-                            })
-                            .catch( error => {
-                                error => {
-                                    onTransition(new FailedSubscriptionState(error));
-                                }
-                            });
-                    }
-                    fetchTokenAndExecuteSubscription();
-                }
-                
-                unsubscribe(){
-                    if(this.tokenPromise) this.tokenPromise.cancel();
-                    this.underlyingSubscription.unsubscribe();
-                    this.onTransition(new EndedSubscriptionState());
-                }
-
-                private putTokenIntoHeader(token: any) {
-                    if(token) {
-                        headers['Authorization'] = `Bearer ${token}`;
-                        logger.verbose(`TokenProvidingSubscription: token fetched: ${token}`);
-                    }
-                }               
-            }
+export let createTokenProvidingStrategy: (
+  nextSubscribeStrategy: SubscribeStrategy,
+  logger: Logger,
+  tokenProvider?: TokenProvider
+) => SubscribeStrategy = (nextSubscribeStrategy, logger, tokenProvider) => {
+  // Token provider might not be provided. If missing, go straight to underlying subscribe strategy.
+  if (tokenProvider) {
+    return (listeners, headers) =>
+      new TokenProvidingSubscription(
+        logger,
+        listeners,
+        headers,
+        tokenProvider,
+        nextSubscribeStrategy
+      );
+  } else {
+    return (listeners, headers) =>
+      nextSubscribeStrategy(listeners, headers);
```

Comment:
You can just return `nextSubscribeStrategy` without wrapping it in an extra function.